### ISSUE TITLE
pl.when().then().otherwise() raises for excluded cases

### ISSUE THREAD

**Issue by m-legrand** [pola-rs/polars#10728]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
def parse_percent(colname: str) -> pl.Expr:
    col = pl.col(colname)
    return (
        pl.when(col.str.ends_with("%"))
        .then(0.01 * col.str.rstrip("%").cast(pl.Float64))
        .otherwise(col.cast(pl.Float64))
    )

pl.DataFrame({"x": ["0.02%"]}).with_columns(parse_percent("x"))
```

### Issue description

```
---------------------------------------------------------------------------
ComputeError                              Traceback (most recent call last)
Cell In[5], line 9
      2     col = pl.col(colname)
      3     return (
      4         pl.when(col.str.ends_with("%"))
      5         .then(0.01 * col.str.rstrip("%").cast(pl.Float64))
      6         .otherwise(col.cast(pl.Float64))
      7     )
----> 9 pl.DataFrame({"x": ["0.02%"]}).with_columns(parse_percent("x"))
[...]
ComputeError: strict conversion from `str` to `f64` failed for value(s) ["0.02%"]; if you were trying to cast Utf8 to temporal dtypes, consider using `strptime`
```

### Expected behavior

`.cast(...)` is only called when the string:
- doesn't end with "%"
- got its final "%" stripped

So it shouldn't fail because of the final "%" character.
I would expect the above example to return a dataframe containing `{"x": [0.0002]}`

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.15
Index type:          UInt32
Platform:            Windows-10-10.0.19045-SP0
Python:              3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          0.3.2a7
deltalake:           <not installed>
fsspec:              2023.6.0
matplotlib:          3.7.2
numpy:               1.25.2
pandas:              2.0.3
pyarrow:             12.0.1
pydantic:            <not installed>
sqlalchemy:          2.0.20
xlsx2csv:            0.8.1
xlsxwriter:          3.1.2
```

</details>

**Comment by mcrumiller** [comment]
This is actually a duplicate of <insert hundreds of issues here>.

See the [warning section in the `pl.when` documentation](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.when.html):

> **Warning**
>
> Polars computes all expressions passed to when-then-otherwise in parallel and filters afterwards. This means each expression must be valid on its own, regardless of the conditions in the when-then-otherwise chain.

All of your `when/then` branches operate on the entire array, which is why you're seeing this issue.

You're better off simply running `.str.rstrip()` on everything without the branch. After all, it won't do anything on the single-branch:

```python
def parse_percent(colname: str) -> pl.Expr:
    col = pl.col(colname)
    return (0.01 * col.str.strip("%").cast(pl.Float64))

pl.DataFrame({"x": ["0.02%"]}).with_columns(parse_percent("x"))
```

**Comment by ritchie46** [comment]
See @mcrumiller detailed explanation. :+1:

**Comment by cmdlineluser** [comment]
If the multiplication is conditional, you could `pl.coalesce` / `.fill_null`
```python
df = pl.DataFrame({"x": ["0.1", "0.02%"]})

df.with_columns(
   pl.coalesce(
      pl.col("x").cast(float, strict=False),
      pl.col("x").str.rstrip("%").cast(float) * 1e-2
   )
)

# shape: (2, 1)
# ┌────────┐
# │ x      │
# │ ---    │
# │ f64    │
# ╞════════╡
# │ 0.1    │
# │ 0.0002 │
# └────────┘
```
